vassal-app/src/main/java/bsh/BeanShellExpressionValidator.java

```diff
@@ -38,7 +40,54 @@
   protected String error;
   protected boolean valid;
 
-  protected final String SUPPORTED_STRING_METHODS = ".length,.contains,.startsWith,.endsWith,.matches,.indexOf,.lastIndexOf,.substring,.replace,";
+  /**
+   * List of all Java 15 String functions that can be applied to property names directly within Beanshell.
+   * (i.e. those with return values and argument types that Vassal supports).
+   *
+   * This list is used to recognize and evaluate constructs of the form
+   *
+   *    stringVariable.function()
+   *
+   * And allow stringVariable to be used directly within beanshell with Java syntax,
+   * rather than having to resort to
+   *
+   *    GetProperty("stringVariable").function()
+   *
+   */
+  private static final Set<String> supportedStringFunctions;
+  static {
+    supportedStringFunctions = new HashSet<>();
+    supportedStringFunctions.add(".compareTo");
+    supportedStringFunctions.add(".compareToIgnoreCase");
+    supportedStringFunctions.add(".contains");
+    supportedStringFunctions.add(".endsWith");
+    supportedStringFunctions.add(".equals");
+    supportedStringFunctions.add(".equalsIgnoreCase");
+    supportedStringFunctions.add(".format");
+    supportedStringFunctions.add(".formatted");
```

Comment:
`formatted` was introduced in Java 15, but we still support Java 11.

Reply:
This list is invisible to users, it does not enable or provide access to these functions, or display them on a menu. It just allows a user to use them as 'native' functions on property names.  I am not adding these functions to the Function Builder menu, but if a smarty-pants designer uses one in their design, Beanshell will not fail with the exception reported by Brian. If they use 'formatted' or 'translateEscapes', they will get a different Beanshell error on a pre-Java 15 system, but that's their problem.

Reply:
The problem is that they'll appear to work for someone designing on Java 15, but then someone else using Java 11 will get exceptions.

Reply:
Ok, I will remove those two from the list.